File: app/src/main/java/com/abishov/hexocat/home/trending/TrendingPresenter.java
package com.abishov.hexocat.home.trending;

import com.abishov.hexocat.commons.network.github.SearchQuery;
import com.abishov.hexocat.commons.schedulers.SchedulerProvider;
import com.abishov.hexocat.commons.utils.OnErrorHandler;
import com.abishov.hexocat.home.repository.RepositoryViewModel;

import io.reactivex.Observable;
import io.reactivex.disposables.CompositeDisposable;

final class TrendingPresenter implements TrendingContract.Presenter {
    private final SchedulerProvider schedulerProvider;
    private final TrendingRepository trendingRepository;
    private final CompositeDisposable compositeDisposable;

    TrendingPresenter(SchedulerProvider schedulerProvider,
            TrendingRepository trendingRepository) {
        this.schedulerProvider = schedulerProvider;
        this.trendingRepository = trendingRepository;
        this.compositeDisposable = new CompositeDisposable();
    }

    @Override
    public void onAttach(TrendingContract.View view) {
        compositeDisposable.add(view.fetchRepositories()
                .switchMap(this::fetchRepositories)
                .subscribe(view.renderRepositories(), OnErrorHandler.create()));
    }

    @Override
    public void onDetach() {
        compositeDisposable.clear();
    }

    private Observable<TrendingViewState> fetchRepositories(SearchQuery query) {
        return trendingRepository.trendingRepositories(query)
                .subscribeOn(schedulerProvider.io())
                .switchMap(repositories -> Observable.fromIterable(repositories)
                        .map(repo -> {
                            String description = repo.description() == null ? "" : repo.description();
                            String forks = String.valueOf(repo.forks());
                            String stars = String.valueOf(repo.stars());
                            return RepositoryViewModel.create(repo.name(), description,
                                    forks, stars, repo.owner().avatarUrl(), repo.owner().login());
                        })
                        .toList().toObservable())
                .map(TrendingViewState::success)
                .startWith(TrendingViewState.progress())
                .onErrorReturn(TrendingViewState::failure)
                .observeOn(schedulerProvider.ui());
    }
}

File: app/src/main/java/com/abishov/hexocat/home/trending/TrendingPresenter.kt
package com.abishov.hexocat.home.trending

import com.abishov.hexocat.common.dagger.FragmentScope
import com.abishov.hexocat.common.schedulers.SchedulerProvider
import com.abishov.hexocat.common.utils.OnErrorHandler
import com.abishov.hexocat.components.*
import com.abishov.hexocat.github.filters.SearchQuery
import com.github.TrendingRepositoriesQuery
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import javax.inject.Inject

private fun TrendingRepositoriesQuery.AsRepository.toViewModel(): RepositoryViewModel {
  val primaryLanguage = primaryLanguage?.let {
    LanguageViewModel(it.name, it.color)
  }

  val topics = repositoryTopics.topics?.mapNotNull { node ->
    node?.let { TopicViewModel(it.topic.name) }
  }

  val contributors = mentionableUsers.contributors?.mapNotNull { node ->
    node?.let { ContributorViewModel(node.id, node.avatarUrl) }
  }

  val users = contributors?.let {
    MentionableUsersViewModel(it, mentionableUsers.totalCount)
  }

  val owner = OwnerViewModel(
    id = owner.id,
    login = owner.login,
    avatarUrl = owner.avatarUrl
  )

  val bannerUrl = if (usesCustomOpenGraphImage) openGraphImageUrl else null

  return RepositoryViewModel(
    name = name,
    description = description,
    bannerUrl = bannerUrl,
    stars = stargazerCount,
    url = url,
    owner = owner,
    primaryLanguage = primaryLanguage,
    topics = topics,
    mentionableUsers = users,
  )
}

@FragmentScope
internal class TrendingPresenter @Inject constructor(
  private val schedulerProvider: SchedulerProvider,
  private val trendingRepositoriesService: TrendingService
) : TrendingContract.Presenter {
  private val compositeDisposable = CompositeDisposable()

  override fun onAttach(view: TrendingContract.View) {
    compositeDisposable.add(view.searchQueries()
      .switchMap { fetchRepositories(it) }
      .subscribe(view.bindTo(), OnErrorHandler.create()))
  }

  override fun onDetach() {
    compositeDisposable.clear()
  }

  private fun fetchRepositories(query: SearchQuery): Observable<TrendingViewState> {
    return trendingRepositoriesService.search(query, 32)
      .subscribeOn(schedulerProvider.io())
      .map { repositories -> repositories.map { it.toViewModel() } }
      .map { TrendingViewState.Success(it) as TrendingViewState }
      .startWith(TrendingViewState.InProgress)
      .onErrorReturn { TrendingViewState.Failure(it.message) }
      .observeOn(schedulerProvider.ui())
  }
}
